fastchat/serve/model_worker.py

```diff
@@ -207,6 +266,36 @@ async def api_generate_stream(request: Request):
     return StreamingResponse(generator, background=background_tasks)
 
 
+@app.post("/worker_generate_completion")
+async def api_generate_completion(request: Request):
+    global model_semaphore, global_counter
+    global_counter += 1
+    params = await request.json()
+
+    if model_semaphore is None:
+        model_semaphore = asyncio.Semaphore(args.limit_model_concurrency)
+    await model_semaphore.acquire()
+    completion = worker.generate_completion(params)
+    background_tasks = BackgroundTasks()
+    background_tasks.add_task(release_model_semaphore)
```

Comment:
duplicated code in `api_generate_completion`, `api_generate_embeddings` and `api_generate_stream`. Try to extract some common utilities for them.